src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
+            if (self.fd == -1) return;
+            const envelope = self.send_queue.out orelse return;
+            self.message_bus.io.send(
+                *ReplicaConnection,
+                self,
+                complete_send,
+                &self.completion,
+                self.fd,
+                envelope.message.buffer[self.bytes_sent..],
+                os.MSG_NOSIGNAL,
+            );
+        }
+
+        fn complete_send(self: *ReplicaConnection, completion: *IO.Completion, result: SendError!usize) void {
+            self.bytes_sent += result catch |err| {
+                // TODO: maybe don't need to close on *every* error
+                log.err("error sending message to replica at {}: {}", .{ self.address, err });
+                self.close();
+                return;
+            };
+            assert(self.bytes_sent <= self.send_queue.out.?.message.buffer.len);
+            // If the message has been fully sent, move on to the next one.
+            if (self.bytes_sent == self.send_queue.out.?.message.buffer.len) {
+                self.bytes_sent = 0;
+                const envelope = self.send_queue.pop().?;
+                envelope.message.references -= 1;
+                self.message_bus.gc(envelope.message);
+                self.message_bus.allocator.destroy(envelope);
+            }
+            self.send();
+        }
+    };
+
+    /// Connection used to recive data from clients and replicas as well as
+    /// send data to clients. This connection is not re-created if it drops,
+    /// The other replica or client is responsible for re-connecting.
+    const Connection = struct {
+        message_bus: *MessageBus,
+        /// A value of -1 indicates that the connection is currently being
+        /// closed or is inactive.
+        fd: os.socket_t = -1,
+        /// The peer is determined by inspecting the first message header
+        /// received.
+        peer: union(enum) {
+            /// No peer is currently connected.
+            none: void,
+            /// A connection has been established but the first header has not yet been recieved.
+            unknown: void,
+            /// The peer is a client with the given id.
+            client: u128,
+            /// The peer is a replica with the given id.
+            replica: u16,
+        } = .none,
+
+        /// This completion is used for all recv operations.
+        recv_completion: IO.Completion = undefined,
+        /// Number of bytes of the current header/message that have already been recieved.
+        bytes_received: usize = 0,
+        incoming_header: Header = undefined,
+        incoming_message: *Message = undefined,
+
+        /// This completion is used for all send operations.
+        send_completion: IO.Completion = undefined,
+        /// The queue of messages to send to the client.
+        /// Empty unless peer == .client
+        send_queue: FIFO(Envelope) = .{},
+        bytes_sent: usize = 0,
```

Comment:
`bytes_received` and `bytes_sent` could be something like `recv_message_size` or `send_message_size` to line up in the source with the syscalls they relate to, to show they compare with `message.header.size` and also to show that they are per message and not running totals (although the comment for `bytes_received` does make that clear). They could both also be ordered in the struct directly after the respective syscall `_completion`. It makes sense for `send_queue` to be last since that's everything waiting on the current `send()`.